dask/dataframe/methods.py

```diff
@@ -304,7 +305,7 @@ def assign(df, *pairs):
     # (to avoid modifying the original)
     pairs = dict(partition(2, pairs))
     deep = bool(set(pairs) & set(df.columns))
-    df = df.copy(deep=bool(deep))
+    df = df.copy() if isinstance(df, dd.core.DataFrame) else df.copy(deep=bool(deep))
```

Comment:
I think we can make this a more general check by using [`dask.utils.has_keyword`](https://github.com/dask/dask/blob/c59a509fad5bb5a7fbe536eb2d593803b8212a22/dask/utils.py#L1557). For example, something like (note I've not tested any of the code below):

```python
if has_keyword(df.copy, "deep"):
    df = df.copy(deep=bool(set(pairs) & set(df.columns)))
else:
    df = df.copy()
```

Reply:
ok, updated - thank you! I specifically encountered this error when I was working with a `<class 'dask_cudf.core.DataFrame'>` - would it be appropriate to add a test case with this type of object, for example in `test_dataframe.py`?

Reply:
Yeah, that sounds great. As a heads up, there are a couple of additional things you'll want to do

1. Use the `@pytest.mark.gpu` pytest marker to make sure the test is run in our gpu-enabled CI build
2. Use `dask_cudf = pytest.importorskip("dask_cudf")` to import `dask_cudf` inside the test to make sure the test is skipped when `dask_cudf` isn't installed (since `dask_cudf` is an optional dependency)

Here's an example of a test which does these two things:

https://github.com/dask/dask/blob/f5b4a82eaf6923a0af01123583ace18a5867c92d/dask/dataframe/tests/test_groupby.py#L2150-L2183

Reply:
@jrbourbeau is it allowed the practice to have a pytest validating in form of a `dask-sql` code?

Reply:
@galipremsagar and I further investigated implementing a test case using `dask-sql` (since that's how we discovered the bug in the first place) but found that those test cases are based on other fixes we've made in Dask-SQL that can't be merged yet, so we can't develop a test case with that either. What would you recommend @jrbourbeau ?

Reply:
We should definitely try to isolate this fix outside the context of `dask-sql`. Grepping through the codebase, it looks like the `assign` function we're updating here is only used in `DataFrame.assign` here

https://github.com/dask/dask/blob/f583b6456db92e22e9dc9236b2ccc36057703877/dask/dataframe/core.py#L4328

A good next step would be to try and trigger the `deep` copy issue using `DataFrame.assign`. We have some existing tests for `DataFrame.assign`. For example

https://github.com/dask/dask/blob/f583b6456db92e22e9dc9236b2ccc36057703877/dask/dataframe/tests/test_dataframe.py#L1399-L1454

Could we slightly modify an existing test to use `dask_cudf` which will trigger the `deep` copy issue you're seeing in `dask-sql`?

Reply:
Right, but I believe the way the issue is being generated in `dask-sql` is that somehow a `<class 'dask_cudf.core.DataFrame'>` object is inserted within another Dask DataFrame, and I'm not really sure how to generate that case. For example, something like: 
```
import cudf, dask_cudf
import dask.dataframe as dd

df = cudf.DataFrame(
    {"a": range(8), "b": [float(i) for i in range(10, 18)]}
) 
ddf = dask_cudf.from_cudf(df, npartitions=1) 

res = ddf.assign( 
 c=1, 
 d="string", 
 e=ddf.a.sum(), 
 f=ddf.a + ddf.b, 
 g=lambda x: x.a + x.b, 
)
```
could already pass without a fix, but I can't figure out how to insert another Dask DataFrame into `ddf`.

Reply:
Hmm that's interesting. For some reason I was thinking the issue was stemming from `cudf` partitions. I'm not sure we should expect having a `dask_cudf` partition to work (cc @quasiben). @nils-braun any thoughts on if this is ultimately an issue with `dask-sql` and not `dask` itself?

Reply:
For further context, I don't think it's necessarily an issue with `dask-sql`, since it works when I use a Pandas DataFrame, just fails when I use a cuDF DataFrame.

Nils can correct me if I'm wrong, but when I run an example query with Dask-SQL like `cc.sql("SELECT id FROM test WHERE EXISTS (SELECT val FROM test WHERE val > 2)").compute()`, the result of the nested select is inserted as an additional DataFrame within a DataFrame. The workflow operates fine from my end, it's just that when Dask-SQL + a Dask-cuDF DataFrame attempts to do a `copy` inside of an `assign`, it can't assign the `deep` parameter. 

Here is the full Dask-SQL + cuDF example: 
```
from dask_sql import Context
import cudf, dask_cudf

cc = Context()

df = cudf.DataFrame()
df['id'] = [0, 1, 1]
df['val'] = [0, 1, 2]

cc.create_table('test', dask_cudf.from_cudf(df, npartitions=2))
cc.sql("SELECT id FROM test WHERE EXISTS (SELECT val FROM test WHERE val > 2)").compute()
```
as opposed to the equivalent Dask-SQL + Pandas example: 
```
from dask_sql import Context
import pandas as pd

cp = Context()

df = pd.DataFrame()
df['id'] = [0, 1, 1]
df['val'] = [0, 1, 2]

cp.create_table('test', df)
cp.sql("SELECT id FROM test WHERE EXISTS (SELECT val FROM test WHERE val > 2)").compute()
```

Reply:
Hi @jrbourbeau and @sarahyurick! Thanks for bringing this to my attention. 
First, to your question on the additional dataframe. Depends on what you mean with "additional" :-) As everything is assumed to be immutable in dask-sql, a new dataframe is created at every step of the relational algebra (in all cases). 

Your shown SQL translates to (you can check with `cp.explain`):
```
LogicalProject(id=[$0])
  LogicalJoin(condition=[true], joinType=[inner])
    LogicalProject(id=[$0])
      LogicalTableScan(table=[[root, test]])
    LogicalProject($f0=[$0])
      LogicalFilter(condition=[IS NOT NULL($0)])
        LogicalAggregate(group=[{}], agg#0=[MIN($0)])
          LogicalProject($f0=[true])
            LogicalFilter(condition=[>($1, 0)])
              LogicalTableScan(table=[[root, test]])
```
so unfortunately its a bit complicated... If the `assign` is the problem (what I assume from your changes ;-)), there are `assigns` happening in the `Join`, the `Aggregate` and the `Project`s. Can you find out which `assign` causes the problem? I am now guessing (I can not test right now), but I *think* it is in the `Join`, specifically in [this](https://github.com/nils-braun/dask-sql/blob/main/dask_sql/physical/rel/logical/join.py#L118). This function ´merge_single_partitions` is only used on these "strange" join conditions (like "true" here) and will be used in a `map_partition` a few lines after that - and here comes the "problem": this `assign` in the function is not an assign on a dask dataframe, but on a partition of a dask dataframe (which is either a pandas dataframe or in your case a cudf dataframe). That might be the problem here? Is it possible to do a `df.assign(common=1)` if `df` is just a normal cudf dataframe?
(if that is not the problem, we have more `assign`s to test ;-))

Reply:
Thanks for your response! I checked the `merge_single_partitions()` function and the issue doesn't appear to be there... I'll keep poking around.

Reply:
I am having trouble wrapping my head around a case where we would need to make a copy of a pd/cudf DataFrame or dd/dask_cudf Dataframe (which should both support the `deep` arg as far as I can tell).

Also, the use of `.compute().compute()` in the motivating example has me a bit worried.  It implies that we are generating a collection of Dask collections, or a collection of `Delayed` objects.  This *may* be what we need, but given that it is not required for the pandas-backed case, we should avoid that can of worms if possible.